Modularize and refactor webpack config

rmainwork opened the pull request:
This PR cleans up the webpack config a little and tries to consolidate configuration into it's most appropriate silo. For example, all client related bundling config now lives in `src/client/webpack.config.js`. This helps to provide webpack config files that are individually runnable for compiling pieces of the project, as well as modularizing the webpack configuration and promoting re-use of the config as much as possible.

## Types of changes
<!--
  What types of changes does your code introduce? Put an `x` in all the boxes
  that apply:
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality
      to change)

## Checklist:
<!--
  Go over all the following points, and put an `x` in all the boxes that apply.
-->
- [ ] I have run `eslint` on the code
- [x] I have added JSDoc for all of my code (where applicable)
- [ ] I have added tests to cover my changes.

## Priority:
- [x] Normal <!-- New piece of functionality -->
- [ ] High <!-- Critical bug requiring urgent review -->

## Related Issues:
<!--
  Add the issue number in below that this PR is intended to address.

  Also mention any issues that this PR is related to, and if possible,
  provide more information regarding the relationship to the issues in
  question (i.e does the PR fix the issue, is it designed to fix another bug
  that is similar to the issue etc.)
-->
Fixes #___

<!--
  Attribution:
  PR Template adapted from: https://github.com/h5bp/html5-boilerplate/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

Commit:
Basic webpack setup

Create basic client and server webpack config files along with top level
webpack config file to import and run both of them.

Both the client and the server have their own independently runnable
webpack config file, which contains config specific for that context.
This can then be imported into the top level config (and overridden if
required) as well as into a webpack dev server config file so that the
config to compile the app is the same as the config that runs it. This
results in a much cleaner webpack setup, as well as allowing more
complex things to be re-used and/or passed down like environment
variables.

Commit:
Webpack Dev Server

Add webpack dev server and inherit server and client config for
consistency

Commit:
Remove `publicPath` because it breaks the path the dev server root path

Setting `publicPath` like this caused webpack-dev-server to try to serve
the app from `/static/`, which is not exactly desirable. So this was
removed

Commit:
Remove old webpack configuration file as it is no longer needed

Commit:
Adjust variable and comment nomenclature

As well as making the language more inclusive, it's also a little more accurate
as to what the code is doing (i.e: ignoring the packages)

Commit:
Update webpack dev server comment block

jonseitz commented on src/client/webpack.config.js:
```diff
@@ -0,0 +1,97 @@
+ * @param {object} webpackEnv Webpack env object (basically any/all options passed in via the CLI)
+ * @param {object} [processEnv] Process env object (environment variables from process.env)
+ */
+const config = ({ mode = 'none' }, processEnv = {}) => ({
```
What's the rationale for making `processEnv` a parameter instead of using `process.env` directly? Because all things being equal I'd prefer to just use `process.env` and have fewer things to (potentially) configure.

robertmain replied:
So, this was meant to match the way that webpack expects you to export functional configs. According to [the documentation](https://v4.webpack.js.org/configuration/configuration-types/#exporting-a-function) the function recieves two arguments from webpack: `env` and `argv`. Passing the value down like that (rather than just using `process.env` directly) helps us to have a better flow of data so there's just one single source of config data.

In answering this question though you made me realize that actually the arguments have somehow gotten flipped, meaning that `mode` was _always_ defaulting to `'development'`. So I've fixed that in e37526ff8245cb027a0ba1ca0565125de28cf1cc

rmainwork replied:
So, this was meant to match the way that webpack expects you to export functional configs. According to [the documentation](https://v4.webpack.js.org/configuration/configuration-types/#exporting-a-function) the function recieves two arguments from webpack: `env` and `argv`. Passing the value down like that (rather than just using `process.env` directly) helps us to have a better flow of data so there's just one single source of config data.

In answering this question though you made me realize that actually the arguments have somehow gotten flipped, meaning that `mode` was _always_ defaulting to `'development'`. So I've fixed that in e37526ff8245cb027a0ba1ca0565125de28cf1cc

jonseitz commented on src/client/webpack.config.js:
```diff
@@ -0,0 +1,97 @@
+    path: paths.dist.app,
+    filename: 'app.js',
+  },
+  optimization: {
+    minimize: true,
+    minimizer: [new TerserPlugin({
+      sourceMap: true,
+    })],
+    noEmitOnErrors: true,
+    nodeEnv: 'production',
+    occurrenceOrder: true,
+    providedExports: true,
+    usedExports: true,
+    sideEffects: true,
+    splitChunks: {
+      chunks: 'all',
+      name: true,
+      minSize: 1,
+      minChunks: 1,
+      cacheGroups: {
+        assets: {
+          test: /[\\/]node_modules[\\/]/,
+          enforce: true,
+          reuseExistingChunk: true,
+          filename: '[name][hash].js',
+        },
+      },
+    },
+  },
```
This problem exists in `master` as well, but the `optimization` block here is specifically for production bundling. W probably want some kind of condition inclusion based on the value of `mode`.

(This has been worked around in the course-planner repo, but we haven't backported that here yet.)

rmainwork replied:
Yes, I did notice this. The primary purpose of this PR was to do a drop-in(ish) replacement for what's currently there and basically just re-organize what we had (with some non-breaking changes). Maybe this is something that could be changed in a future PR?

jonseitz commented on package.json:
```diff
@@ -37,7 +37,7 @@
     "test:watch": "TS_NODE_TRANSPILE_ONLY=true mocha --watch",
     "typecheck": "tsc --noEmit",
-    "start": "nodemon"
+    "start": "nodemon & webpack-dev-server --config webpack.dev-server.config --config-name=client"
```
So this older config is actually running webpack's hot middle in our Nest server (see: `src/server/middleware/dev.middleware.ts` and its application in `src/server/app.module.ts`). So running both `nodemon` and the standalone `webpack-dev-server` will create two hot-reloading.

When this was created the goal was to have both client and server running in a single process. Instead we probably want to switch to using separate start scripts for the client and server (e.g. `start:client` and `start:server`) so we have finer-grained control over the two (and run them in separate docker containers)

robertmain replied:
Ah, I'd actually forgotten about that when I included webpack-dev-server! Given that (and what I said above about not changing things around too much) it might make sense to remove this (for this PR anyway)?

rmainwork replied:
Ah, good point. I'd forgotten we'd done it that way. In that case, I won't bring in a second instance of `webpack-dev-server`.

jonseitz commented on docker-compose.yml:
```diff
@@ -32,6 +32,7 @@ services:
     user: "1000:1000"
     ports:
       - "127.0.0.1:3000:3000"
+      - "127.0.0.1:8000:8000"
```
Another thing we need to backport from course-planner: These should use the same SERVER_PORT and CLIENT_PORT environment values that are being passing into the container

jonseitz commented on webpack.dev-server.config.js:
```diff
@@ -0,0 +1,69 @@
+          { from: /./, to: '/index.html' },
+        ],
+      },
+      port: 8000,
```
I think we want to get the `CLIENT_PORT` value from the environment here instead

jonseitz commented on webpack.dev-server.config.js:
```diff
@@ -0,0 +1,69 @@
+*/
+module.exports = ({ mode = 'development' } = {}) => ([
+  {
+    ...client.config({ mode }, process.env),
```
The dev-server also needs some extra config under `resolve`:

```js
    alias: {
      'react-dom': '@hot-loader/react-dom',
    },
```

as in https://github.com/seas-computing/react-boilerplate/blob/master/src/server/middleware/dev.middleware.ts#L49-L51.

As well as the `react-hot-loader` plugin under `modules`:

```js
        use: [
          {
            loader: 'babel-loader',
            options: {
              plugins: ['react-hot-loader/babel'],
            },
          },
          'ts-loader',
```

https://github.com/seas-computing/react-boilerplate/blob/master/src/server/middleware/dev.middleware.ts#L35-42

robertmain replied:
Whoops! I must've missed those when I copied the client config over...

jonseitz commented on src/server/webpack.config.js:
```diff
@@ -0,0 +1,120 @@
+    sideEffects: true,
+  },
+  module: {
+    rules: [
```
These aren't in the boilerplate, but we should backport the loaders for css and other generic files from the course-planner repo:

```js
      {
        test: /\.css$/i,
        use: ['style-loader', 'css-loader'],
      },
      {
        test: /\.(woff(2)?|ttf|eot|svg)(\?v=\d+\.\d+\.\d+)?$/,
        use: [
          {
            loader: 'file-loader',
            options: {
              name: '[name].[ext]',
              outputPath: 'assets/',
            },
          },
        ],
      },
```

jonseitz commented:
[Whoops, sent the review without adding a message]

Overall this looks good and I appreciate the modular approach. On a somewhat semantic note, I don't know if the `src/` directory is quite the right place to have these config files, since they're not _really_ source code. And it leads to the slighty awkward `resolve(__dirname, '..', '..', 'build') call to get the root directory. What if we just had them in root-level `webpack.client.config.js` and `webpack.server.config.js` files instead? 


There is also the problem that we're missing a lot of the more up-to-date webpack configurations from the course-planner project. I've noted a few of them here, but even beyond webpack I think we've made some significant changes in how we build and develop the app:

1. We're running completely separate processes for the client and server, in separate containers.
2. We're no longer user `webpack` for production bundling of our server code, but instead just using `tsc` to transpile everything. We are still using it for development so we can get hot module reloading within nest.

I think both of those changes are going to stick around for future project, which means we don't really need separate prod/dev server webpack configs, just server ones, and we no longer have the use case of needing to combine both into a single production config.

It may more sense to do this in a separate PR, but I think we should try to backport more of the new course-planner configs into the boilerplate and refactor based on _that_ code, rather than refactor the old code and then backport the changes on there. Does that seem fair?

Commit:
Fix webpack config argument order

It turns out that the webpack config/env argument order had gotten
flipped meaning that any additional webpack arguments or config passed
via the CLI were essentially ignored.
